### PR TITLE
Feature/202102/mizinkobusters/minigame

### PR DESCRIPTION
**Pull request by Mizinkobusters**

### CODE REVIEW

**Comment by Mizinkobusters** [comment]
randomstoneコマンドが正常に作動しなかったため要修正
以下エラーログを添付

`
[07:54:30] [Server thread/ERROR]: null
org.bukkit.command.CommandException: Unhandled exception executing command 'randomstone' in plugin MiningGame v1.0.0
	at org.bukkit.command.PluginCommand.execute(PluginCommand.java:47) ~[spigot-1.16.5.jar:git-Spigot-138d451-b9b0c4e]
	at org.bukkit.command.SimpleCommandMap.dispatch(SimpleCommandMap.java:149) ~[spigot-1.16.5.jar:git-Spigot-138d451-b9b0c4e]
	at org.bukkit.craftbukkit.v1_16_R3.CraftServer.dispatchCommand(CraftServer.java:762) ~[spigot-1.16.5.jar:git-Spigot-138d451-b9b0c4e]
	at net.minecraft.server.v1_16_R3.PlayerConnection.handleCommand(PlayerConnection.java:1788) ~[spigot-1.16.5.jar:git-Spigot-138d451-b9b0c4e]
	at net.minecraft.server.v1_16_R3.PlayerConnection.c(PlayerConnection.java:1631) ~[spigot-1.16.5.jar:git-Spigot-138d451-b9b0c4e]
	at net.minecraft.server.v1_16_R3.PlayerConnection.a(PlayerConnection.java:1584) ~[spigot-1.16.5.jar:git-Spigot-138d451-b9b0c4e]
	at net.minecraft.server.v1_16_R3.PacketPlayInChat.a(PacketPlayInChat.java:47) ~[spigot-1.16.5.jar:git-Spigot-138d451-b9b0c4e]
	at net.minecraft.server.v1_16_R3.PacketPlayInChat.a(PacketPlayInChat.java:1) ~[spigot-1.16.5.jar:git-Spigot-138d451-b9b0c4e]
	at net.minecraft.server.v1_16_R3.PlayerConnectionUtils.lambda$0(PlayerConnectionUtils.java:19) ~[spigot-1.16.5.jar:git-Spigot-138d451-b9b0c4e]
	at net.minecraft.server.v1_16_R3.TickTask.run(SourceFile:18) ~[spigot-1.16.5.jar:git-Spigot-138d451-b9b0c4e]
	at net.minecraft.server.v1_16_R3.IAsyncTaskHandler.executeTask(SourceFile:144) ~[spigot-1.16.5.jar:git-Spigot-138d451-b9b0c4e]
	at net.minecraft.server.v1_16_R3.IAsyncTaskHandlerReentrant.executeTask(SourceFile:23) ~[spigot-1.16.5.jar:git-Spigot-138d451-b9b0c4e]
	at net.minecraft.server.v1_16_R3.IAsyncTaskHandler.executeNext(SourceFile:118) ~[spigot-1.16.5.jar:git-Spigot-138d451-b9b0c4e]
	at net.minecraft.server.v1_16_R3.MinecraftServer.bb(MinecraftServer.java:942) ~[spigot-1.16.5.jar:git-Spigot-138d451-b9b0c4e]
	at net.minecraft.server.v1_16_R3.MinecraftServer.executeNext(MinecraftServer.java:935) ~[spigot-1.16.5.jar:git-Spigot-138d451-b9b0c4e]
	at net.minecraft.server.v1_16_R3.IAsyncTaskHandler.awaitTasks(SourceFile:127) ~[spigot-1.16.5.jar:git-Spigot-138d451-b9b0c4e]
	at net.minecraft.server.v1_16_R3.MinecraftServer.sleepForTick(MinecraftServer.java:919) ~[spigot-1.16.5.jar:git-Spigot-138d451-b9b0c4e]
	at net.minecraft.server.v1_16_R3.MinecraftServer.w(MinecraftServer.java:851) ~[spigot-1.16.5.jar:git-Spigot-138d451-b9b0c4e]
	at net.minecraft.server.v1_16_R3.MinecraftServer.lambda$0(MinecraftServer.java:164) ~[spigot-1.16.5.jar:git-Spigot-138d451-b9b0c4e]
	at java.lang.Thread.run(Unknown Source) [?:1.8.0_281]
Caused by: java.lang.OutOfMemoryError: Java heap space
	at java.util.Arrays.copyOf(Unknown Source) ~[?:1.8.0_281]
	at java.util.ArrayList.grow(Unknown Source) ~[?:1.8.0_281]
	at java.util.ArrayList.ensureExplicitCapacity(Unknown Source) ~[?:1.8.0_281]
	at java.util.ArrayList.ensureCapacityInternal(Unknown Source) ~[?:1.8.0_281]
	at java.util.ArrayList.add(Unknown Source) ~[?:1.8.0_281]
	at mb.mizinkobusters.elysium.command.Randomstone.getBlocks(Randomstone.java:158) ~[?:?]
	at mb.mizinkobusters.elysium.command.Randomstone.generateBlocks(Randomstone.java:89) ~[?:?]
	at mb.mizinkobusters.elysium.command.Randomstone.onCommand(Randomstone.java:60) ~[?:?]
	at org.bukkit.command.PluginCommand.execute(PluginCommand.java:45) ~[spigot-1.16.5.jar:git-Spigot-138d451-b9b0c4e]
	at org.bukkit.command.SimpleCommandMap.dispatch(SimpleCommandMap.java:149) ~[spigot-1.16.5.jar:git-Spigot-138d451-b9b0c4e]
	at org.bukkit.craftbukkit.v1_16_R3.CraftServer.dispatchCommand(CraftServer.java:762) ~[spigot-1.16.5.jar:git-Spigot-138d451-b9b0c4e]
	at net.minecraft.server.v1_16_R3.PlayerConnection.handleCommand(PlayerConnection.java:1788) [spigot-1.16.5.jar:git-Spigot-138d451-b9b0c4e]
	at net.minecraft.server.v1_16_R3.PlayerConnection.c(PlayerConnection.java:1631) [spigot-1.16.5.jar:git-Spigot-138d451-b9b0c4e]
	at net.minecraft.server.v1_16_R3.PlayerConnection.a(PlayerConnection.java:1584) [spigot-1.16.5.jar:git-Spigot-138d451-b9b0c4e]
	at net.minecraft.server.v1_16_R3.PacketPlayInChat.a(PacketPlayInChat.java:47) [spigot-1.16.5.jar:git-Spigot-138d451-b9b0c4e]
	at net.minecraft.server.v1_16_R3.PacketPlayInChat.a(PacketPlayInChat.java:1) [spigot-1.16.5.jar:git-Spigot-138d451-b9b0c4e]
	at net.minecraft.server.v1_16_R3.PlayerConnectionUtils.lambda$0(PlayerConnectionUtils.java:19) [spigot-1.16.5.jar:git-Spigot-138d451-b9b0c4e]
	at net.minecraft.server.v1_16_R3.PlayerConnectionUtils$$Lambda$4171/1997757153.run(Unknown Source) [spigot-1.16.5.jar:git-Spigot-138d451-b9b0c4e]
	at net.minecraft.server.v1_16_R3.TickTask.run(SourceFile:18) [spigot-1.16.5.jar:git-Spigot-138d451-b9b0c4e]
	at net.minecraft.server.v1_16_R3.IAsyncTaskHandler.executeTask(SourceFile:144) [spigot-1.16.5.jar:git-Spigot-138d451-b9b0c4e]
	at net.minecraft.server.v1_16_R3.IAsyncTaskHandlerReentrant.executeTask(SourceFile:23) [spigot-1.16.5.jar:git-Spigot-138d451-b9b0c4e]
	at net.minecraft.server.v1_16_R3.IAsyncTaskHandler.executeNext(SourceFile:118) [spigot-1.16.5.jar:git-Spigot-138d451-b9b0c4e]
	at net.minecraft.server.v1_16_R3.MinecraftServer.bb(MinecraftServer.java:942) [spigot-1.16.5.jar:git-Spigot-138d451-b9b0c4e]
	at net.minecraft.server.v1_16_R3.MinecraftServer.executeNext(MinecraftServer.java:935) [spigot-1.16.5.jar:git-Spigot-138d451-b9b0c4e]
	at net.minecraft.server.v1_16_R3.IAsyncTaskHandler.awaitTasks(SourceFile:127) [spigot-1.16.5.jar:git-Spigot-138d451-b9b0c4e]
	at net.minecraft.server.v1_16_R3.MinecraftServer.sleepForTick(MinecraftServer.java:919) [spigot-1.16.5.jar:git-Spigot-138d451-b9b0c4e]
	at net.minecraft.server.v1_16_R3.MinecraftServer.w(MinecraftServer.java:851) [spigot-1.16.5.jar:git-Spigot-138d451-b9b0c4e]
	at net.minecraft.server.v1_16_R3.MinecraftServer.lambda$0(MinecraftServer.java:164) [spigot-1.16.5.jar:git-Spigot-138d451-b9b0c4e]
	at net.minecraft.server.v1_16_R3.MinecraftServer$$Lambda$3204/1678956565.run(Unknown Source) [spigot-1.16.5.jar:git-Spigot-138d451-b9b0c4e]
	... 1 more
[07:54:30] [Server thread/WARN]: Can't keep up! Is the server overloaded? Running 9910ms or 198 ticks behind
`

File: src/mb/mizinkobusters/elysium/command/Randomstone.java [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+        }
+
+        Player player = (Player) sender;
+
+        if (!player.isOp()) {
+
+            player.sendMessage("§c権限がありません");
+            return true;
+
+        }
```
Comment (themadpillow) [inline src/mb/mizinkobusters/elysium/command/Randomstone.java:48]: よく考えたらplugin.ymlで権限指定してたらコマンド入力時点で弾かれるのでここ無くても大丈夫ですね
（xprankingのコマンドも含め）

File: src/mb/mizinkobusters/elysium/command/Randomstone.java [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+                public void run() {
+
+                    getBlocks().iterator().next().getLocation().getBlock().setType(getGenerateMaterial().getType());
+                    Bukkit.getLogger().info("Placed a block.");
+
+                }
+
+            }.runTaskTimer(plugin, 20,5);
+
+        }
```
Comment (themadpillow) [inline src/mb/mizinkobusters/elysium/command/Randomstone.java:102]: - Thread内でgetBlocks()毎回生成してるように見えるけどこれ終わる？
- whileの位置的にブロック分のThreadが作られてしまう気がする
- 置くたびにログ出すと表示がとんでもない気がするので終わったタイミングでのログ表示にしたい

File: src/config.yml [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+  yaw: 0
+  pitch: 0
+
+# /randomstoneを実行した際に起点となる場所を設定します.
+# 設定した地点からX, Z軸の正の向きにブロックが生成されます.
+setOriginBlock:
+  world: "world"
+  x: 100
+  y: 100
+  z: 100
```
Comment (themadpillow) [inline src/config.yml:16]: 起点のポイントはconfigより開始前にコマンドで設定か、/startコマンド入力地点の方が良いかも
ここから変更するのは面倒だったら後に別PRでもOK

File: src/mb/mizinkobusters/elysium/command/Randomstone.java [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+        if (random.nextInt(result_lapisLazuli) == 0) {
+            return new ItemStack(Material.LAPIS_ORE);
+        }
+
+        int result_redstone = 100 / ratio_redstone;
+        if (random.nextInt(result_redstone) == 0) {
+            return new ItemStack(Material.REDSTONE_ORE);
+        }
+
+        return new ItemStack(Material.STONE);
```
Comment (themadpillow) [inline src/mb/mizinkobusters/elysium/command/Randomstone.java:241]: - random.nextInt(Int)は 「0~(引数-1)までの数値をランダムで取得」するので == 0だとかなり厳しい判定になるのでは
- この判定方法でも間違ってないと思うけど、毎回乱数生成するのはコスト高そうなので一度0~100の乱数生成してその値に応じてそれぞれの鉱石に振り分けたいかも（となると石の比率も入力必要かな？）

File: src/mb/mizinkobusters/elysium/command/Start.java [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+        }
+
+        Player player = (Player) sender;
+
+        if (!player.isOp()) {
+
+            player.sendMessage("§c権限がありません");
+            return true;
+
+        }
```
Comment (themadpillow) [inline src/mb/mizinkobusters/elysium/command/Start.java:53]: これも他コマンドと同様になくても良い、可能なものはplugin.ymlで管理したい

**Review by themadpillow — COMMENTED** [review]
途中ですが一旦あげときます
また後で残り見ます

File: src/mb/mizinkobusters/elysium/command/Start.java [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+
+    private Player getHighestLeveler() {
+
+        List<Player> sortedPlayers = Bukkit.getOnlinePlayers().stream()
+                .sorted(Comparator.comparingInt(Player::getExpToLevel))
+                .collect(Collectors.toList());
+
+        return Bukkit.getPlayer(sortedPlayers.get(0).getPlayer().getUniqueId());
+
+    }
```
Comment (themadpillow) [inline src/mb/mizinkobusters/elysium/command/Start.java:211]: getHighestLeveler()は他のコマンドでも見た気がするので何度も使われるようなものならUtilクラスなどで切り分けてそっちを利用するようにした方がいいですね

File: src/mb/mizinkobusters/elysium/command/Start.java [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+
+            }
+        }.runTaskTimer(plugin, 0, 20);
+
+        new BukkitRunnable() {
+
+            @Override
+            public void run() {
+
+                Bukkit.dispatchCommand(player, "xpranking");
```
Comment (themadpillow) [inline src/mb/mizinkobusters/elysium/command/Start.java:115]: PluginからPlayerにコマンドを実行させるとそのPlayerがオフラインだった場合などエラーになりそうな気がするのと、
回りくどい感じがするのでxprankingコマンドを入力したときの処理を直接呼ぶなどの対応が必要な気がします
他のBukkit#dispatchCommandを使用している部分も同様

**Review by themadpillow — COMMENTED** [review]
好みの範疇かなと思ってましたが、全体通して空行が目立つのでメソッド内の前後やif文内などの前後は空行は無い方が良い気がします 🤔 

https://kazurof.github.io/GoogleJavaStyle-ja/#s4-6-whitespace
参考までにGoogleのJavaコーディングスタイルガイド（上記は非公式日本語訳）的には空行使用箇所は主に以下の通りです。
1. クラスの連続するメンバか初期化子の間。フィールド、コンストラクタ、メソッド、ネストしたクラス、static初期化子、インスタンス初期化子
1. ステートメントの間で、コードを論理的にグループ分けしたいため
1.  必要な場合、クラスの最初のメンバーあるいは初期化子の前と最終メンバーあるいは初期化子の後

例：
```
class Sample() {
                                                // 空行（任意[3]）
  String a = "a";
                                                // 空行（必須[1]、例外として変数が2個のみの場合は任意）
  String b = "b";
                                                // 空行（必須[1] 任意[3]でもある）
  public static void main(String[] args) {
    String a1 = add1String(a);
    String b1 = add1String(b);
                                                // 空行（必須[2] 変数の初期化と処理のグループ分け）
    if (args.length() != 0 && args[0].equals("show")) {
      System.out.println(a1 + b1);
    }
    return;
  }
                                                // 空行（必須[1]）
  public String add1String(String string) {
    return string + "1";
  }
}
```

File: src/mb/mizinkobusters/elysium/command/RandomStone.java [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+        }
+
+        player.sendMessage("§a生成を開始しています...");
+        player.sendMessage("§7この操作には時間がかかる場合があります");
+        generateBlocks();
+        return true;
+    }
+
+    private void generateBlocks() {
+        FileConfiguration config = plugin.getConfig();
```
Comment (themadpillow) [inline src/mb/mizinkobusters/elysium/command/RandomStone.java:53]: 毎回configのバリデーションチェックをしている関係で可読性落ちている感じがするので、
configのチェックを行う別クラスを作成してデータも一つのクラスで管理してそこから利用する方が良いかもです

File: src/mb/mizinkobusters/elysium/command/Start.java [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+                    this.cancel();
+                    return;
+                }
+
+                if (i == 0) {
+                    for (Player players : Bukkit.getOnlinePlayers()) {
+                        players.sendTitle("スタート!", "", 5, 100, 5);
+                        players.playSound(players.getLocation(), Sound.ENTITY_GENERIC_EXPLODE, 1, 1);
+                    }
+                    return;
```
Comment (themadpillow) [inline src/mb/mizinkobusters/elysium/command/Start.java:78]: スタート表示したらcancelしてしまっていいのでここまとめられそうですね

File: src/mb/mizinkobusters/elysium/command/Start.java [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+                for (Player players : Bukkit.getOnlinePlayers()) {
+                    if (0 < i) {
+                        players.sendMessage(i + "...");
+                        players.playSound(players.getLocation(), Sound.BLOCK_ANVIL_FALL, 1, 1);
+                    }
+                }
+                i--;
+                n--;
+            }
+        }.runTaskTimer(plugin, 35900, 20);
```
Comment (themadpillow) [inline src/mb/mizinkobusters/elysium/command/Start.java:145]: 35900が何秒（分）なのか一見分からないので（意図的に36000ではなくしてる？）
30 * 60 * 20などするか、別途毎秒回るTaskを作ってまとめても良いかもですね（どのみちTimer表示もしたい感じするので）

File: src/mb/mizinkobusters/elysium/command/Start.java [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+            players.addPotionEffect(new PotionEffect(PotionEffectType.GLOWING, 1000000, 127, false, false));
+            players.addPotionEffect(new PotionEffect(PotionEffectType.DAMAGE_RESISTANCE, 1000000, 127, false, false));
+        }
+
+        GenerateBlockThread thread = new GenerateBlockThread(plugin);
+        thread.generate();
+
+        new BukkitRunnable() {
+            int timer = 0;
+            int i = 0;
```
Comment (themadpillow) [inline src/mb/mizinkobusters/elysium/command/Start.java:66]: 変数iが何を表しているのかわかりにくいかもです

File: src/mb/mizinkobusters/elysium/command/Start.java [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+            @Override
+            public void run() {
+                if (60 * 30 + 5 < timer) {
+                    this.cancel();
+                    return;
+                }
+
+                if (timer == 1) {
+                    i = 3;
+                }
```
Comment (themadpillow) [inline src/mb/mizinkobusters/elysium/command/Start.java:77]: timerは 60*30+5 以下にはならないように設定されていますが、
timerが1になった場合の処理があります。ここは確実に通らない気がします。

他のTimerの値判定部分も同様

File: src/mb/mizinkobusters/elysium/GenerateBlockThread.java [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+
+    public GenerateBlockThread(JavaPlugin plugin) {
+        this.plugin = plugin;
+    }
+
+    public void generate() {
+        FileConfiguration config = plugin.getConfig();
+        if (!config.contains("setOriginBlock.world") || !config.isString("setOriginBlock.world")) {
+            return;
+        }
```
Comment (themadpillow) [inline src/mb/mizinkobusters/elysium/GenerateBlockThread.java:28]: Configのバリデーションチェックを行う場所は別がいいかもです
Configのみを管理するクラスなどを作りたいですね

File: src/mb/mizinkobusters/elysium/command/Start.java (hunk 11 above)
Comment (themadpillow) [reply src/mb/mizinkobusters/elysium/command/Start.java:60]: このTimer関係だけ修正されれば残りはOKだと思います。

File: src/mb/mizinkobusters/elysium/command/Start.java (hunk 12 above)
Comment (themadpillow) [reply src/mb/mizinkobusters/elysium/command/Start.java:71]: このTimer関係だけ修正されれば残りはOKだと思います。

**Review by themadpillow — APPROVED** [review]
PRが肥大化してきたので細かい部分は一旦置きます LGTMです！ 🚀

**Pull request merged**